crates/adapters/src/transport/mod.rs

```diff
@@ -157,21 +248,56 @@ pub trait InputEndpoint: Send {
 ///
 /// A transport endpoint pushes binary data downstream via an instance of this
 /// trait.
-// TODO: `input_owned`.
+///
+/// For a durable endpoint, where the data is divided into steps, there is some
+/// special terminology:
+///
+///   * "Committed" steps.  A step is "committed" when the endpoint has added
+///     all of the data to it that it is going to.  The reader indicates that a
+///     step `step`, and all prior steps, are committed by calling
+///     `InputConsumer::start_step(step + 1)`.
+///
+///     A committed step may not yet be durable.  Committing indicates that the
+///     endpoint writing it to stable storage, but that might not be complete
+///     yet.  The controller can start processing the input step but it should
+///     not yet yield any side effects that can't be retracted.
+///
+///   * "Settled" steps.  This is the term for a committed step that has been
+///     written to stable storage.  The reader indicates that a step `step`, and
+///     all prior steps, have settled by calling `InputConsumer::settled(step)`.
+///
+///     Settled is really a synonym for "durable", but that term is already used
+///     too much in this API.
 pub trait InputConsumer: Send {
+    /// Indicates that upcoming calls are for `step`.
+    fn start_step(&mut self, step: Step);
+
     /// Push a fragment of the input stream to the consumer.
     ///
     /// `data` is not guaranteed to start or end on a record boundary.
     /// The parser is responsible for identifying record boundaries and
     /// buffering incomplete records to get prepended to the next
     /// input fragment.
+    ///
+    /// A durable input transport keeps the order of fragments the same for a
+    /// given step from one read to the next.
     fn input_fragment(&mut self, data: &[u8]) -> Vec<ParseError>;
 
     /// Push a chunk of data to the consumer.
     ///
     /// The chunk is expected to contain complete records only.
+    ///
+    /// A durable input transport might shuffle the order of chunks for a given
```

Comment:
Hmm, that guarantee is already there, I just need to rewrite the comment, like this:

```rust
    /// Push a chunk of data to the consumer.
    ///
    /// The chunk is expected to contain complete records only.
    ///
    /// Some data in a durable input transport might not have an inherently
    /// defined order within a step.  The input endpoint may shuffle unordered
    /// chunks within a step from one read to the next.  For example, the
    /// durable Kafka reader will provide chunks from a given partition in the
    /// same order on each read, but it might interleave chunks from different
    /// partitions differently each time.
    fn input_chunk(&mut self, data: &[u8]) -> Vec<ParseError>;
```